File: libraries/cyclestreets-core/src/main/java/net/cyclestreets/api/client/dto/PoiTypesDto.java
package net.cyclestreets.api.client.dto;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;

import com.fasterxml.jackson.annotation.JsonProperty;

import net.cyclestreets.api.POICategories;
import net.cyclestreets.api.POICategory;
import net.cyclestreets.util.Base64;

import java.io.ByteArrayInputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class PoiTypesDto {
  @JsonProperty(value = "validuntil")
  private long validUntil;
  @JsonProperty
  private Map<String, PoiTypeDto> types;

  public static class PoiTypeDto {
    @JsonProperty
    private String id;
    @JsonProperty
    private String name;
    @JsonProperty
    private String total;
    @JsonProperty
    private String icon;

    private POICategory toPOICategory(Context context) {
      return new POICategory(id, name, poiIcon(context, icon));
    }
  }

  public POICategories toPOICategories(Context context) {
    List<POICategory> categories = new ArrayList<>();
    for (Map.Entry<String, PoiTypeDto> entry : types.entrySet()) {
      categories.add(entry.getValue().toPOICategory(context));
    }
    return new POICategories(categories);
  }

  private static Drawable poiIcon(final Context context,
                                  final String iconAsBase64) {
    final Bitmap bmp = decodeIcon(iconAsBase64);
    return new BitmapDrawable(context.getResources(), bmp);
  }

  private static Bitmap decodeIcon(final String iconAsBase64) {
    try {
      byte[] bytes = Base64.decode(iconAsBase64);
      return BitmapFactory.decodeStream(new ByteArrayInputStream(bytes));
    } catch (Exception e) {
      return null;
    }
  }
}

File: libraries/cyclestreets-core/src/main/java/net/cyclestreets/api/client/dto/PoiTypesDto.kt
package net.cyclestreets.api.client.dto

import android.content.Context
import android.graphics.drawable.Drawable
import androidx.core.content.res.ResourcesCompat
import com.fasterxml.jackson.annotation.JsonProperty
import net.cyclestreets.api.POICategories
import net.cyclestreets.api.POICategory
import net.cyclestreets.core.R


class PoiTypesDto {
    @JsonProperty(value = "validuntil")
    private val validUntil: Long = 0

    @JsonProperty
    private val types: Map<String?, PoiTypeDto>? = null

    class PoiTypeDto {
        @JsonProperty
        private lateinit var id: String

        @JsonProperty
        private lateinit var name: String

        @JsonProperty
        private lateinit var total: String

        fun toPOICategory(context: Context): POICategory {
            return POICategory(id, name, poiIcon(context, id))
        }
    }

    fun toPOICategories(context: Context): POICategories {
        val categories: MutableList<POICategory> = ArrayList()
        for ((_, value) in types!!) {
            categories.add(value.toPOICategory(context))
        }
        return POICategories(categories)
    }

}


private fun poiIcon(context: Context, id: String): Drawable {
    val key = "poi_${id}"
    val res = context.resources
    val resPackage = res.getResourcePackageName(defaultResId)

    val resId = res.getIdentifier(key, "drawable", resPackage)
    val resIdOrDefault = if (resId != 0) resId else defaultResId

    return ResourcesCompat.getDrawable(res, resIdOrDefault, null)!!
}

private val defaultResId = R.drawable.poi_attractions
